### PR TITLE
ITP Pipeline Setup

### PR DESCRIPTION
**Pull request by tpan496**
- This adds a way to take in an ITP packet, convert it into either a `ReplicationCommand` or `StopReplicationCommand` and pass it to the traffic cop. Right now there is no distinction between the implementation of commands, as it just involves sending packets to the traffic cop.
- We can build a `ReplicationCommand` using the `ITPPacketWriter`.
- This PR also involves some refactoring for `ProtocolInterpreter`.

### CODE REVIEW

File: src/include/storage/recovery/replication_log_provider.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+#include <memory>
+#include "storage/recovery/abstract_log_provider.h"
+
+namespace terrier::storage {
+
+/**
+ * TODO(Tianlei/Gus): Replace this file with Gus's version
+ */
+class ReplicationLogProvider {
```
Comment (tpan496) [inline src/include/storage/recovery/replication_log_provider.h]: @GustavoAngulo This is just a basic frame I created. I am not sure if this is sufficient.

File: src/network/postgres/postgres_protocol_interpreter.cpp [hunk 2]
```diff
@@ -74,54 +74,11 @@ Transition PostgresProtocolInterpreter::ProcessStartup(const std::shared_ptr<Rea
   return Transition::PROCEED;
 }
 
-bool PostgresProtocolInterpreter::TryBuildPacket(const std::shared_ptr<ReadBuffer> &in) {
-  if (!TryReadPacketHeader(in)) return false;
+size_t PostgresProtocolInterpreter::GetPacketHeaderSize() { return startup_ ? sizeof(int32_t) : 1 + sizeof(int32_t); }
```
Comment (GustavoAngulo) [inline src/network/postgres/postgres_protocol_interpreter.cpp]: These should be `uint32_t`

File: src/include/network/itp/itp_packet_writer.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  /**
+   * End the Replication command
+   */
+  void EndReplicationCommand() { EndPacket(); }
+
+  /**
+   * Writes a Stop Replication packet
+   */
+  void StopReplicationCommand() { BeginPacket(NetworkMessageType::ITP_STOP_REPLICATION_COMMAND); }
```
Comment (YuzeLiao) [inline src/include/network/itp/itp_packet_writer.h]: Do we need to invoke EndPacket() here?

File: src/network/itp/itp_protocol_interpreter.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  if (command->FlushOnComplete()) out->ForceFlush();
+  Transition ret = command->Exec(common::ManagedPointer<ProtocolInterpreter>(this),
+                                 common::ManagedPointer<ITPPacketWriter>(&writer), t_cop, context, callback);
+  curr_input_packet_.Clear();
+  return ret;
+}
+
+void ITPProtocolInterpreter::GetResult(std::shared_ptr<WriteQueue> out) {
+  ITPPacketWriter writer(out);
+  writer.BeginPacket(NetworkMessageType::ITP_COMMAND_COMPLETE).EndPacket();
```
Comment (YuzeLiao) [inline src/network/itp/itp_protocol_interpreter.cpp]: It seems we can directly invoke WriteCommandComplete()？

File: src/include/network/itp/itp_protocol_interpreter.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  /**
+   * @see ProtocolIntepreter::Process
+   * @param in
+   * @param out
+   * @param callback
+   * @param t_cop the traffic cop pointer
+   * @param context the connection context
+   * @return
+   */
```
[the remark this answers is not in the document]
Comment (tpan496) [reply src/include/network/itp/itp_protocol_interpreter.h]: The comments would likely be redundant though. Is `@see ProtocolIntepreter::Process` not sufficient?

File: src/include/network/itp/itp_protocol_interpreter.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  /**
+   * Writes result to the client
+   * @param out
+   */
+  void GetResult(std::shared_ptr<WriteQueue> out) override;
+
+ protected:
+  /**
+   * @see ProtocolInterpreter::GetPacketHeaderSize
+   * Header format: 5 byte message size
```
[the remark this answers is not in the document]
Comment (tpan496) [reply src/include/network/itp/itp_protocol_interpreter.h]: updated

File: src/network/itp/itp_network_commands.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+Transition StopReplicationCommand::Exec(common::ManagedPointer<ProtocolInterpreter> interpreter,
+                                        common::ManagedPointer<ITPPacketWriter> out,
+                                        common::ManagedPointer<trafficcop::TrafficCop> t_cop,
+                                        common::ManagedPointer<ConnectionContext> connection,
+                                        NetworkCallback callback) {
+  // TODO(Tianlei): modify this implementation for the complete ITP protocol
+  std::unique_ptr<ReadBuffer> buffer;
+  buffer->FillBufferFrom(in_, in_len_);
+  t_cop->HandBufferToReplication(std::move(buffer));
```
[the remark this answers is not in the document]
Comment (tpan496) [reply src/network/itp/itp_network_commands.cpp]: Updated

File: src/network/itp/itp_protocol_interpreter.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+#include <string>
+#include <utility>
+
+#include "network/itp/itp_network_commands.h"
+#include "network/itp/itp_protocol_interpreter.h"
+#include "network/network_defs.h"
+#include "network/terrier_server.h"
+
+#define SSL_MESSAGE_VERNO 80877103
+#define PROTO_MAJOR_VERSION(x) ((x) >> 16)
```
[the remark this answers is not in the document]
Comment (tpan496) [reply src/network/itp/itp_protocol_interpreter.cpp]: Updated

File: src/network/itp/itp_protocol_interpreter.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  curr_input_packet_->Clear();
+  return ret;
+}
+
+void ITPProtocolInterpreter::GetResult(std::shared_ptr<WriteQueue> out) {
+  ITPPacketWriter writer(out);
+  writer.BeginPacket(NetworkMessageType::ITP_COMMAND_COMPLETE).EndPacket();
+}
+
+size_t ITPProtocolInterpreter::GetPacketHeaderSize() { return 1 + sizeof(int32_t); }
```
[the remark this answers is not in the document]
Comment (tpan496) [reply src/network/itp/itp_protocol_interpreter.cpp]: Updated

File: src/network/postgres/postgres_protocol_interpreter.cpp (hunk 2 above)
Comment (tpan496) [reply src/network/postgres/postgres_protocol_interpreter.cpp]: Updated

File: src/include/network/itp/itp_packet_writer.h (hunk 3 above)
Comment (tpan496) [reply src/include/network/itp/itp_packet_writer.h]: Yes, I don't think we need to append anything in this case.

**Comment by mbutrovich** [comment]
I'll merge this once it passes CI.

**Pull request merged**